Mock cores/nodes cleanup

markscheel opened the pull request:
## Proposed changes

Cleans up some of the ActionTesting code enabling mocking cores and nodes,
following comments made by @kidder and @nilsdeppe from PR #2792

### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->

<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `major new feature` if appropriate.

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

markscheel committed:
Parallel/Info.hpp: Make all functions noexcept.

markscheel committed:
MockRuntimeSystem: Make dox clearer.

markscheel committed:
ActionTesting: Fix dox typo.

markscheel committed:
Revert "Invoke.hpp: Allow a proxy to disable ckLocal."

This reverts commit 5a6431af55c36762452797753b0f2c022eb5d164.

ckLocal_enabled is no longer needed.

markscheel committed:
ActionTesting: replace ckLocal_enabled by a Requires.

nilsvu commented on tests/Unit/Framework/MockRuntimeSystem.hpp:
```diff
@@ -176,25 +176,23 @@ class MockRuntimeSystem {
-    for (const auto& node_and_core : mock_nodes_and_local_cores_) {
-      const size_t global_core =
-          mock_global_cores_.at(node_and_core.first).at(node_and_core.second);
+    for (const auto& [node,local_core] : mock_nodes_and_local_cores_) {
```
(optional nitpick) run clang-format over this (missing space)

nilsvu commented on tests/Unit/Framework/MockRuntimeSystem.hpp:
```diff
@@ -176,25 +176,23 @@ class MockRuntimeSystem {
           mutable_caches_.at(global_core).get());
       tmpl::for_each<typename Metavariables::component_list>(
-          [this, &global_core, &node_and_core](auto component) {
+          [this, &global_core, &node, &local_core](auto component) {
```
I don't understand why this works with all compilers. Structures bindings are somehow different to variables and can't be captured.

nilsvu commented on tests/Unit/Framework/MockDistributedObject.hpp:
```diff
@@ -129,6 +138,54 @@ struct LocalCoreId {
+  return not(lhs==rhs);
+}
+
+/// Wraps a size_t representing the global core number.
```
Could you add a few words here and to `LocalCoreId` explaining their difference? It's reasonably obvious (I assume "local" identifies the core in the node and "global" identifies the core among all nodes), but it would be good to say that in the dox. Generally, I don't like docstrings that just repeat the symbol name without CamelCase 😄

nilsvu commented on tests/Unit/Framework/MockDistributedObject.hpp:
```diff
@@ -129,6 +138,54 @@ struct LocalCoreId {
+  return lhs.value == rhs.value;
+}
+
+inline bool operator!=(const GlobalCoreId& lhs,
```
Why inline? I never know when to mark functions inline actually..

markscheel replied:
I'm not sure when they should be inlined either.  I did it here because other examples I looked at were inlined.

wthrowe replied:
Roughly, `inline` allows the definition of a non-template function in a header file without the linker complaining.  Functions defined in a class body are implicitly inline.

nilsvu commented on tests/Unit/Framework/MockRuntimeSystem.hpp:
```diff
@@ -176,14 +184,19 @@ class MockRuntimeSystem {
+      // Note that lambdas cannot capture structured bindings,
+      // so we define node and local_core here.
+      const auto& node = node_and_local_core.first.value;
+      const auto& local_core = node_and_local_core.second.value;
```
Ah this explains my previous comment :) Probably better to fix the earlier commit so it compiles.

markscheel committed:
ActionTesting: lists of nodes/cores are more typesafe.

Previously the lists of nodes and cores were things like
std::vector<std::pair<size_t,size_t>>
where local cores, global cores, and nodes were all represented
by a size_t.

Now inside the lists, we use types GlobalCoreId, LocalCoreId,
and NodeId so that it is always clear what all the pieces of
the lists mean.

markscheel committed:
MockRuntimeSystem: Use structured bindings.

markscheel commented:
Made changes (except didn't do anything about inlining).  Squashed, because some of the changes involved moving things between commits anyway.

nilsdeppe reviewed (approved):
LGTM, please give @kidder a chance to take a look
